### PR TITLE
feature(config): add support for static config modifications with extending dynamic config

### PR DESCRIPTION
**Pull request by byCedric**
# Why

This adds support in `@expo/config` to modify static configs, when used with extending dynamic configs. This only applies to static & dynamic configs written as:

- **app.json** → `{ ... }` (any json content)
- **app.config.js** → `({ config }) => ({ ...config, version: 'x.x.x' })` (extending static config)

# How

- Added test cases to ensure configs are loaded properly
- Added support for static + function-like dynamic config
- Added validation to ensure the static config results in eventual config modifications, by reloading config after change

> ⚠️ Note, currently we have to manually merge the dynamic config result as only static configs are returned from `modifyConfigAsync`. In a follow-up PR, I'll simplify this, dropping the `getIntersectingObject` helper to skip merging.
> See: https://github.com/expo/expo/pull/30783

# Test Plan

See CI

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by byCedric** [comment]
* **#30782** <a href="https://app.graphite.dev/github/pr/expo/expo/30782?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @byCedric and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/@expo/config/src/__tests__/Config-test.ts [hunk 1]
```diff
@@ -306,3 +306,120 @@ describe('hasUnusedStaticConfig', () => {
...
+  const appFile = {
+    name: 'testing 123',
+    version: '0.1.0',
+    slug: 'testing-123',
+    sdkVersion: '100.0.0',
+  };
+
+  afterEach(() => vol.reset());
+
+  xit('succeeds without existing config', async () => {
```
Comment (byCedric) [inline packages/@expo/config/src/__tests__/Config-test.ts:346]:
```suggestion
  it('succeeds without existing config', async () => {
```

File: packages/@expo/config/src/__tests__/Config-test.ts [hunk 2]
```diff
@@ -306,3 +306,120 @@ describe('hasUnusedStaticConfig', () => {
...
+    createProject('/no-config');
+
+    await expect(modifyConfigAsync('/no-config', { name: 'new name' })).resolves.toMatchObject({
+      type: 'success',
+      config: { name: 'new name' },
+    });
+  });
+
+  xit('succeeds when modifying static config only', async () => {
```
Comment (byCedric) [inline packages/@expo/config/src/__tests__/Config-test.ts:355]:
```suggestion
  it('succeeds when modifying static config only', async () => {
```

File: packages/@expo/config/src/__tests__/Config-test.ts [hunk 3]
```diff
@@ -306,3 +306,120 @@ describe('hasUnusedStaticConfig', () => {
...
+      'app.json': JSON.stringify(appFile, null, 2),
+    });
+
+    await expect(modifyConfigAsync('/static-only', { name: 'new name' })).resolves.toMatchObject({
+      type: 'success',
+      config: { ...appFile, name: 'new name' },
+    });
+  });
+
+  xit('warns when modifying dynamic config only', async () => {
```
Comment (byCedric) [inline packages/@expo/config/src/__tests__/Config-test.ts:366]:
```suggestion
  it('warns when modifying dynamic config only', async () => {
```

File: packages/@expo/config/src/__tests__/Config-test.ts [hunk 4]
```diff
@@ -306,3 +306,120 @@ describe('hasUnusedStaticConfig', () => {
...
+    });
+
+    await expect(modifyConfigAsync('/dynamic-only', { name: 'new name' })).resolves.toMatchObject({
+      type: 'warn',
+      config: null,
+      message: expect.stringMatching(/dynamic config/),
+    });
+  });
+
+  xit('warns when modifying static with object-like dynamic config', async () => {
```
Comment (byCedric) [inline packages/@expo/config/src/__tests__/Config-test.ts:378]:
```suggestion
  it('warns when modifying static with object-like dynamic config', async () => {
```

File: packages/@expo/config/src/__tests__/Config-test.ts [hunk 5]
```diff
@@ -306,3 +306,120 @@ describe('hasUnusedStaticConfig', () => {
...
+    await expect(
+      modifyConfigAsync('/static-dynamic-object', { name: 'new name' })
+    ).resolves.toMatchObject({
+      type: 'warn',
+      config: null,
+      message: expect.stringMatching(/dynamic config/),
+    });
+  });
+
+  xit('warns when modifying static with function-like dynamic config, without extending config', async () => {
```
Comment (byCedric) [inline packages/@expo/config/src/__tests__/Config-test.ts:393]:
```suggestion
  it('warns when modifying static with function-like dynamic config, without extending config', async () => {
```

**Review by byCedric — COMMENTED** [review]
Re-enabling the new test cases

**Review by brentvatne — APPROVED** [review]
🤩

File: packages/@expo/config/src/__tests__/Config-test.ts [hunk 6]
```diff
@@ -306,3 +306,120 @@ describe('hasUnusedStaticConfig', () => {
...
+      'app.config.js': `module.exports = JSON.parse(\`${JSON.stringify({ ...appFile, version: '9.9.9' })}\`);`,
+    });
+
+    await expect(
+      modifyConfigAsync('/static-dynamic-object', { name: 'new name' })
+    ).resolves.toMatchObject({
+      type: 'warn',
+      config: null,
+      message: expect.stringMatching(/dynamic config/),
+    });
```
Comment (EvanBacon) [inline packages/@expo/config/src/__tests__/Config-test.ts:390]: Missing an assertion that the app.json was rolled back.

File: packages/@expo/config/src/__tests__/Config-test.ts [hunk 7]
```diff
@@ -306,3 +306,120 @@ describe('hasUnusedStaticConfig', () => {
...
+  });
+
+  it('succeeds when modifying static with function-like dynamic config, with extending config', async () => {
+    createProject('/static-dynamic-function', {
+      'app.json': JSON.stringify(appFile),
+      'app.config.js': `module.exports = ({ config }) => ({ ...config, version: '9.9.9' });`,
+    });
+
+    await expect(
+      modifyConfigAsync('/static-dynamic-function', { name: 'new name' })
```
Comment (EvanBacon) [inline packages/@expo/config/src/__tests__/Config-test.ts:415]: Test with a more complex / nested object.

**Review by EvanBacon — APPROVED** [review]
missing two parts for the test, otherwise looks great!

File: packages/@expo/config/src/__tests__/Config-test.ts (hunk 7 above)
Comment (byCedric) [reply packages/@expo/config/src/__tests__/Config-test.ts:415]: @EvanBacon done!

File: packages/@expo/config/src/__tests__/Config-test.ts (hunk 6 above)
Comment (byCedric) [reply packages/@expo/config/src/__tests__/Config-test.ts:390]: Done!

**Pull request merged**